storage/grpc_metrics.go

```diff
@@ -0,0 +1,295 @@
+	}
+	updates := []attribute.KeyValue{}
+	for _, kv := range gcpAttributeExpectedDefaults() {
+		if val, present := s.Value(kv.Key); !present || val.AsString() == "" {
+			updates = append(updates, attribute.KeyValue{Key: kv.Key, Value: kv.Value})
+		}
+	}
+	r, err := resource.New(
+		ctx,
+		resource.WithAttributes(
+			attribute.KeyValue{Key: "gcp.resource_type", Value: attribute.StringValue(monitoredResourceName)},
+			attribute.KeyValue{Key: "instance_id", Value: attribute.StringValue(uuid.New().String())},
+			attribute.KeyValue{Key: "project_id", Value: attribute.StringValue(project)},
+			attribute.KeyValue{Key: "api", Value: attribute.StringValue("grpc")},
+		),
+		resource.WithAttributes(detectedAttrs.Attributes()...),
+		// Last duplicate key / value wins
+		resource.WithAttributes(updates...),
+	)
+	if err != nil {
+		return nil, err
+	}
+	preparedResource.resource = r
+	return preparedResource, nil
+}
+
+type internalMetricsConfig struct {
+	project  string
+	endpoint string
+}
+
+type internalMetricsContext struct {
+	// monitoring API endpoint used
+	endpoint string
+	// project used by exporter
+	project string
+	// client options passed to gRPC channels
+	clientOpts []option.ClientOption
+	// instance of metric reader used by gRPC client-side metrics
+	provider *metric.MeterProvider
+	// clean func to call when closing gRPC client
+	close func()
+}
+
+func (mc *internalMetricsContext) String() string {
+	return fmt.Sprintf("endpoint: %v\nproject: %v\n", mc.endpoint, mc.project)
+}
+
+func determineMonitoringEndpoint(endpoint string) string {
+	// Check storage endpoint in case its using VPC then we use that endpoint instead.
+	if strings.Contains(endpoint, "private.googleapis.com") || strings.Contains(endpoint, "restricted.googleapis.com") {
+		return endpoint
+	}
+	// Default monitoring endpoint is used.
+	return ""
+}
+
+func createHistogramView(name, desc, unit string, boundaries []float64) metric.View {
+	return metric.NewView(metric.Instrument{
+		Name:        name,
+		Description: name,
+		Kind:        metric.InstrumentKindHistogram,
+		Unit:        unit,
+	}, metric.Stream{
+		Name:        name,
+		Description: desc,
+		Unit:        unit,
+		Aggregation: metric.AggregationExplicitBucketHistogram{Boundaries: boundaries},
+	})
+}
+
+func newGRPCMetricContext(ctx context.Context, config internalMetricsConfig) (*internalMetricsContext, error) {
+	preparedResource, err := newPreparedResource(ctx, config.project, []resource.Option{resource.WithDetectors(gcp.NewDetector())})
+	if err != nil {
+		return nil, err
+	}
+	if config.project != preparedResource.projectToUse {
+		log.Printf("The Project ID configured for metrics is %s, but the Project ID of the storage client is %s. Make sure that the service account in use has the required metric writing role (roles/monitoring.metricWriter) in the project projectIdToUse or metrics will not be written.", preparedResource.projectToUse, config.project)
```

Comment:
How would this arise? Also this should probably return an error from this func?

Reply:
Implementation uses OTel Detector to determine project-id to use and if that project is different from the project used by the storage client then this warning is emitted so it's clear which project is selected. If detector doesn't detect a project it will fallback to ADC. However, if no project is found then client-side metrics cannot start. I'll add a case for this.